File: 2019/python/day_04a.py
#!/usr/bin/python3

import sys


def meetsConditions(num):
    prev_val = 10
    repeat = False
    while num > 0:
        val = num % 10
        num = int(num / 10)
        if val > prev_val:
            return False
        if prev_val == val:
            repeat = True
        prev_val = val
    return repeat


def main():
    # Get input
    file_name = "../input/day_04_input"
    if len(sys.argv) > 1:
        file_name = sys.argv[1]
    file = open(file_name)
    val_range = [int(ele) for ele in file.readline().strip("\n").split("-")]

    # Solve
    count = 0
    for num in range(val_range[0], val_range[1] + 1):
        if meetsConditions(num):
            count += 1
    print(count)
    return count


if __name__ == "__main__":
    main()

File: 2019/cpp/day_04a.cpp
#include <fstream>
#include <iostream>
#include <string>

bool meetsConditions(int num) {
  int prev_val = 10;
  bool repeat = false;
  while (num > 0) {
    const auto val = num % 10;
    num = int(num / 10);
    if (val > prev_val) {
      return false;
    }
    if (prev_val == val) {
      repeat = true;
    }
    prev_val = val;
  }
  return repeat;
}

int main(int argc, char *argv[]) {
  // Get input
  std::string input = "../input/day_04_input";
  if (argc > 1) {
    input = argv[1];
  }
  std::ifstream file(input);
  std::string input_str;
  std::getline(file, input_str);

  const auto delimiter_index = input_str.find("-");
  auto min_val = std::stoi(input_str.substr(0, delimiter_index));
  auto max_val = std::stoi(input_str.substr(
      delimiter_index + 1, input_str.size() - delimiter_index - 1));

  // Solve
  size_t count = 0;
  for (int num = min_val; num < max_val; ++num) {
    if (meetsConditions(num)) {
      count += 1;
    }
  }

  std::cout << count << '\n';
  return count;
}
